### PR TITLE
Add testing of bitflips in packet headers

### PR DESCRIPTION
**Pull request by t8m**
A new type of noise is introduced in the noisy dgram bio filter.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
This needs to be rebased after #22243 is merged.

File: test/helpers/noisydgrambio.c [hunk 1]
```diff
@@ -114,6 +118,36 @@ static void get_noise(uint64_t *reinject, int *should_drop)
      * extra datagram in that case
      */
     *reinject += (uint64_t)(*should_drop);
+
+    /* flip some bits in the header */
+    if (type == NOISE_TYPE_BITFLIPS) {
+        /* we flip at most 8 bits of the 16 bit value at once */
+        *flip = (test_random() % 255 + 1) << (test_random() % 8);
+        *flip_offset = test_random() % (25 * (1 + long_header));
```
Comment (mattcaswell) [inline test/helpers/noisydgrambio.c:126]: 25 seems quite a magic number here?

File: test/helpers/noisydgrambio.c [hunk 2]
```diff
@@ -114,6 +118,36 @@ static void get_noise(uint64_t *reinject, int *should_drop)
...
+    }
+}
+
+static void flip_bits(unsigned char *msg, size_t msg_len, uint16_t flip,
+                      size_t flip_offset)
+{
+    if (flip == 0)
+        return;
+
+    /* None of these border conditions should not happen but check them anyway */
```
Comment (mattcaswell) [inline test/helpers/noisydgrambio.c:136]: Typo: None of these border conditions should happen

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
> This needs to be rebased after https://github.com/openssl/openssl/pull/22243 is merged.

Care will need to be taken when rebasing. Flipped bits will (presumably) cause the datagram to be dropped, so we will need to back off that type of noise too.

**Review by paulidale — APPROVED** [review]
With @mattcaswell's suggestions taken.

**Comment by mattcaswell** [comment]
This needs a rebase now that #22243 has been merged.

**Comment by t8m** [comment]
Rebased and @mattcaswell comments addressed.

Please reconfirm.

File: test/helpers/noisydgrambio.c (hunk 1 above)
Comment (t8m) [reply test/helpers/noisydgrambio.c:141]: I've added a comment.

**Review by mattcaswell — DISMISSED** [review]

**Comment by mattcaswell** [comment]
The no-cmp runchecker CI failure is relevant

**Comment by mattcaswell** [comment]
CI failure is still happening

File: test/helpers/quictestlib.c [hunk 3]
```diff
@@ -235,6 +235,7 @@ int qtest_create_quic_objects(OSSL_LIB_CTX *libctx, SSL_CTX *clientctx,
         if (!TEST_ptr(noisebio))
             goto err;
         sbio = BIO_push(noisebio, sbio);
+        (void)BIO_ctrl(sbio, BIO_CTRL_NOISE_BACK_OFF, 0, NULL);
```
[the remark this answers is not in the document]
Comment (t8m) [reply test/helpers/quictestlib.c:238]: Yep

**Review by mattcaswell — APPROVED** [review]
Approved subject to the CI passing this time!

**Comment by t8m** [comment]
> Approved subject to the CI passing this time!

It passed!

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for the reviews.

**Pull request closed without merging**